### PR TITLE
Refactor: instruction account index

### PR DESCRIPTION
**Pull request by Lichtso**
#### Problem
`index_in_instruction` is a mixed index, meaning that it maps both program accounts and instruction accounts in one sequence.

`index_in_instruction = program_account_index`
or
`index_in_instruction = number_of_program_accounts + instruction_account_index`

In most cases however, we only want to use one: Either program accounts or instruction accounts.
Thus, it is better to split their interfaces to avoid conflation and also to reduce the index shifting in the code base.
Also, some interfaces were already split, because they only make sense for one of the two account types,
so this PR makes the interfaces more uniform and hopefully less confusing as well.

#### Summary of Changes

Split methods of `InstructionContext`:
- `get_index_in_transaction()` into `get_index_of_program_account_in_transaction()` and `get_index_of_instruction_account_in_transaction()`
- `try_borrow_account()` into `try_borrow_program_account()` and `try_borrow_instruction_account()`

Renamed methods of `InstructionContext`:
- `try_borrow_program_account()` to `try_borrow_last_program_account()`
- `get_program_key()` to `get_last_program_key()`
- `find_index_of_account()` to `find_index_of_instruction_account()`
- `is_duplicate()` to `is_instruction_account_duplicate()`
- `is_signer()` to `is_instruction_account_signer()`
- `is_writer()` to `is_instruction_account_writer()`

Removed methods of `InstructionContext`:
- `get_program_id_index()`, use `get_index_of_program_account_in_transaction()` instead
- `get_program_id()` use `get_last_program_key()` instead
- `get_number_of_accounts()` as it became unused

Removed methods of `BorrowedAccount`:
- `get_index_in_instruction()` as it became unused

The renaming is on purpose to avoid mixing index basis (e.g. when backporting) by making the function names different / incompatible.

### CODE REVIEW

**Review by jackcmay — APPROVED** [review]

**Comment by jstarry** [comment]
Yes, will take a look today

File: runtime/src/system_instruction_processor.rs [hunk 1]
```diff
@@ -33,7 +33,7 @@ struct Address {
 }
 
 impl Address {
-    fn is_signer(&self, signers: &HashSet<Pubkey>) -> bool {
+    fn is_instruction_account_signer(&self, signers: &HashSet<Pubkey>) -> bool {
```
Comment (jstarry) [inline runtime/src/system_instruction_processor.rs:36]: I don't think this rename is necessary, why did you change it?

File: sdk/src/transaction_context.rs [hunk 2]
```diff
@@ -646,15 +658,31 @@ impl<'a> BorrowedAccount<'a> {
 
     /// Returns whether this account is a signer (instruction wide)
     pub fn is_signer(&self) -> bool {
+        if self.index_in_instruction < self.instruction_context.program_accounts.len() {
+            return false;
+        }
         self.instruction_context
-            .is_signer(self.index_in_instruction)
+            .instruction_accounts
```
Comment (jstarry) [inline sdk/src/transaction_context.rs:665]: You can use `is_instruction_account_signer` here

File: sdk/src/transaction_context.rs [hunk 3]
```diff
@@ -646,15 +658,31 @@ impl<'a> BorrowedAccount<'a> {
...
     }
 
     /// Returns whether this account is writable (instruction wide)
     pub fn is_writable(&self) -> bool {
+        if self.index_in_instruction < self.instruction_context.program_accounts.len() {
+            return false;
+        }
         self.instruction_context
-            .is_writable(self.index_in_instruction)
+            .instruction_accounts
```
Comment (jstarry) [inline sdk/src/transaction_context.rs:680]: Can use `is_instruction_account_writable` here

**Review by jstarry — APPROVED** [review]

File: runtime/src/system_instruction_processor.rs (hunk 1 above)
Comment (Lichtso) [reply runtime/src/system_instruction_processor.rs:36]: Good spotting, was regex replace going to far.
